src/audio_core/hle/wmf_decoder.cpp

```diff
@@ -0,0 +1,254 @@
+        std::memcpy(&response, &request, sizeof(response));
+        response.unknown1 = 0x0;
+        return response;
+    }
+    default:
+        LOG_ERROR(Audio_DSP, "Got unknown binary request: {}", static_cast<u16>(request.cmd));
+        return {};
+    }
+}
+
+std::optional<BinaryResponse> WMFDecoder::Impl::Initalize(const BinaryRequest& request) {
+    if (initalized) {
+        Clear();
+    }
+
+    BinaryResponse response;
+    std::memcpy(&response, &request, sizeof(response));
+    response.unknown1 = 0x0;
+
+    if (mf_decoder_init(&transform) != 0) {
+        LOG_CRITICAL(Audio_DSP, "Can't init decoder");
+        return response;
+    }
+
+    HRESULT hr = transform->GetStreamIDs(1, &in_stream_id, 1, &out_stream_id);
+    if (hr == E_NOTIMPL) {
+        // if not implemented, it means this MFT does not assign stream ID for you
+        in_stream_id = 0;
+        out_stream_id = 0;
+    } else if (FAILED(hr)) {
+        ReportError("Decoder failed to initialize the stream ID", hr);
+        SafeRelease(&transform);
+        return response;
+    }
+
+    initalized = true;
+    return response;
+}
+
+void WMFDecoder::Impl::Clear() {
+    if (initalized) {
+        mf_flush(&transform);
+        mf_deinit(&transform);
+    }
+    initalized = false;
+    selected = false;
+}
+
+int WMFDecoder::Impl::DecodingLoop(ADTSData adts_header,
+                                   std::array<std::vector<u8>, 2>& out_streams) {
+    int output_status = 0;
+    char* output_buffer = NULL;
+    DWORD output_len = 0;
+    IMFSample* output = NULL;
+
+    while (true) {
+        output_status = receive_sample(transform, out_stream_id, &output);
+
+        // 0 -> okay; 3 -> okay but more data available (buffer too small)
+        if (output_status == 0 || output_status == 3) {
+            copy_sample_to_buffer(output, (void**)&output_buffer, &output_len);
+
+            // the following was taken from ffmpeg version of the decoder
+            f32 val_f32;
+            for (size_t i = 0; i < output_len;) {
+                for (std::size_t channel = 0; channel < adts_header.channels; channel++) {
+                    std::memcpy(&val_f32, output_buffer + i, sizeof(val_f32));
+                    s16 val = static_cast<s16>(0x7FFF * val_f32);
+                    out_streams[channel].push_back(val & 0xFF);
+                    out_streams[channel].push_back(val >> 8);
+                    i += sizeof(val_f32);
+                }
+            }
+
+            if (output_buffer)
+                free(output_buffer);
+        }
+
+        // in case of "ok" only, just return quickly
+        if (output_status == 0)
+            return 0;
+
+        // for status = 2, reset MF
+        if (output_status == 2) {
+            Clear();
+            return -1;
+        }
+
+        // for status = 3, try again with new buffer
+        if (output_status == 3)
+            continue;
+
+        return output_status; // return on other status
```

Comment:
LOG_ERROR the output_status in this case?

Reply:
Addressed